oprf/group/group.go

```diff
@@ -0,0 +1,117 @@
+package group
+
+import (
+	"hash"
+	"math/big"
+)
+
+// CiphersuiteID is
+type CiphersuiteID int
+
+const (
+	// P256 is the group generated by P-256 elliptic curve.
+	P256 CiphersuiteID = iota
+)
+
+// Ciphersuite corresponds to the OPRF ciphersuite that is chosen. The
+
+// Ciphersuite corresponds to the OPRF ciphersuite that is chosen. The
+// Ciphersuite object determines the prime-order group (pog) that is used for
+// performing the OPRF operations, along with the different hash function
+// definitions.
+// Should be created using New, using a CiphersuiteID constant.
+type Ciphersuite struct {
+	// name of the ciphersuite
+	name string
+
+	// context
+	ctx string
+
+	// Suite instantiation for performing the OPRF operations.
+	suite Suite
+}
+
+// Suite is an interface that defines operations within additive
+// groups of prime order. This is the setting in which the OPRF operations
+// take place.
+type Suite interface {
+	// Creates a new Suite object
+	New(CiphersuiteID) (Suite, error)
+
+	// Returns the identifying name of the group
+	Name() string
+
+	// Returns the canonical (fixed) generator for defined group
+	Generator() *Point
+
+	// Returns the order of the canonical generator in the group.
+	Order() *big.Int
+
+	// Base hash function used in conjunction with the Ciphersuite
+	Hash() hash.Hash
+
+	// Returns k * G, where G is the canonical generator of the group, and k is
+	// some scalar value provided as input.
+	ScalarMult(*Scalar) (*Point, error)
+
+	// Performs a transformation to encode bytes as a Point object in the
+	// group.
+	HashToGroup([]byte) (*Point, error)
+
+	// Performs a transformation to encode bytes as a Scalar object in the
+	// appropiate group.
+	HashToScalar([]byte) (Scalar, error)
+
+	// Samples a random scalar value from the field of scalars defined by the
+	// group order.
+	RandomScalar() (*Scalar, error)
+
+	// Returns the ByteLength of objects associated with the Ciphersuite
+	ByteLength() int
+}
+
+// Point is the interface that represents points in a given
+// Ciphersuite instantiation.
+type Point interface {
```

Comment:
Should this be `Element`, to better align with pogs?

Reply:
Yes. I'll like it to be aligned with the draft as well. But I haven't seen a cryptographic library that refers to points as elements.. so might be confusing.

Reply:
The fact to named it element is because `Group` is an abstraction from elliptic curves.
When we add other implementations of a `Group`, e.g. decaf, the elements are not exactly points.

Reply:
OK ;) I'll rename it element ;)